kernel/tests/golden_tables.rs

```diff
@@ -0,0 +1,343 @@
+    let temp_dir = tempfile::tempdir()?;
+    archive.unpack(temp_dir.path())?;
+    Ok(temp_dir)
+}
+
+// NB adapated from DAT: read all parquet files in the directory and concatenate them
+async fn read_expected(path: &Path) -> DeltaResult<Option<RecordBatch>> {
+    let store = Arc::new(LocalFileSystem::new_with_prefix(path)?);
+    let files = store.list(None).try_collect::<Vec<_>>().await?;
+    let mut batches = vec![];
+    let mut schema = None;
+    for meta in files.into_iter() {
+        if let Some(ext) = meta.location.extension() {
+            if ext == "parquet" {
+                let reader = ParquetObjectReader::new(store.clone(), meta);
+                let builder = ParquetRecordBatchStreamBuilder::new(reader).await?;
+                if schema.is_none() {
+                    schema = Some(builder.schema().clone());
+                }
+                let mut stream = builder.build()?;
+                while let Some(batch) = stream.next().await {
+                    batches.push(batch?);
+                }
+            }
+        }
+    }
+    let all_data = concat_batches(&schema.unwrap(), &batches)?;
+    Ok(Some(all_data))
+}
+
+// TODO: change to do something similar to dat tests instead of string comparison
+macro_rules! assert_batches_eq {
+    ($expected: expr, $chunks: expr) => {
+        let formatted_expected = arrow::util::pretty::pretty_format_batches(&[$expected])
+            .unwrap()
+            .to_string();
+        let formatted_result = arrow::util::pretty::pretty_format_batches(&[$chunks])
+            .unwrap()
+            .to_string();
+        let mut sorted_expected: Vec<&str> = formatted_expected.trim().lines().collect();
+        let mut sorted_result: Vec<&str> = formatted_result.trim().lines().collect();
+        sort_lines!(sorted_expected);
+        sort_lines!(sorted_result);
+        assert_eq!(
+            sorted_expected, sorted_result,
+            "\n\nexpected:\n\n{:#?}\nactual:\n\n{:#?}\n\n",
+            sorted_expected, sorted_result,
+        );
+    };
+}
+
+// copypasta from DAT
+// some things are equivalent, but don't show up as equivalent for `==`, so we normalize here
+fn normalize_col(col: &Arc<dyn Array>) -> Arc<dyn Array> {
+    if let DataType::Timestamp(unit, Some(zone)) = col.data_type() {
+        if **zone == *"+00:00" {
+            arrow_cast::cast::cast(&col, &DataType::Timestamp(*unit, Some("UTC".into())))
+                .expect("Could not cast to UTC")
+        } else {
+            col.clone()
+        }
+    } else {
+        col.clone()
+    }
+}
+
+// do a full table scan at the latest snapshot of the table and compare with the expected data
+async fn latest_snapshot_test(test_name: &str) -> Result<(), Box<dyn std::error::Error>> {
+    let test_dir = load_test_data(test_name).unwrap();
+    let test_path = test_dir.path().join(test_name);
+    let table_path = std::fs::canonicalize(test_path.join("table")).unwrap();
```

Comment:
can we `expect("Couldn't canonicalize path")` and similar with all the unwraps below. helps with debugging

Reply:
gonna do these in the morning